gh-138425: Correctly partially evaluate global generics with undefined params in `ref.evaluate(format=Format.FORWARDREF)`

dr-carlos opened the pull request:
<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNNN: Summary of the changes made
```

Where: gh-NNNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNNNN)
```

Where: [X.Y] is the branch name, for example: [3.13].

GH-NNNNNN refers to the PR number from `main`.

-->


<!-- gh-issue-number: gh-138425 -->
* Issue: gh-138425
<!-- /gh-issue-number -->

dr-carlos committed:
Consider globals in `ref.evaluate()` for undefined generic `ForwardRef`s

dr-carlos committed:
Test `ref.evaluate()` on undefined generics

dr-carlos committed:
Check that generic is not a ForwardRef before accessing its attrs

dr-carlos committed:
Add NEWS entry

auvipy reviewed (approved):
looks good

sobolevn commented on Misc/NEWS.d/next/Library/2025-09-03-18-26-07.gh-issue-138425.cVE9Ho.rst:
```diff
@@ -0,0 +1,2 @@
+Partially evaluate undefined global generics in
+:py:meth:`annotationlib.ForwardRef.evaluate`
```
```suggestion
:meth:`annotationlib.ForwardRef.evaluate`
```

dr-carlos replied:
Thanks!

sobolevn commented on Lib/test/test_annotationlib.py:
```diff
@@ -1675,6 +1675,24 @@ def test_name_lookup_without_eval(self):
+        generic = ForwardRef("list[int, undef]").evaluate(format=Format.FORWARDREF, globals={"list": dict})
+        self.assertNotIsInstance(generic, ForwardRef)
+        self.assertIs(generic.__origin__, dict)
+        self.assertIs(generic.__args__[0], int)
```
What about other args?

dr-carlos replied:
I didn't think it was necessary to check the other args - I'll check them now.

dr-carlos committed:
Update NEWS Entry

Co-authored-by: sobolevn <mail@sobolevn.me>

JelleZijlstra commented on Misc/NEWS.d/next/Library/2025-09-03-18-26-07.gh-issue-138425.cVE9Ho.rst:
```diff
@@ -0,0 +1,2 @@
+Partially evaluate undefined global generics in
```
Mention of "generics" is imprecise; annotationlib doesn't know anything about generics and the issue affects any names that come from the globals.

dr-carlos replied:
Yes, I really wasn't sure how to describe the issue.
The problem is that this code path is only triggered after two conditions are met:
- The `ForwardRef` arg is not an identifier or keyword, so it's almost certainly a generic (or `GenericAlias`)
- The `ForwardRef` doesn't evaluate, so there needs to be something undefined in it
- And then the issue only arises for names in the globals.

But I'm very happy to change the description here - do you have any suggestions for wording?

JelleZijlstra replied:
It could also be a union `undefined | int`, or an `Annotated` metadata, or just something that isn't a valid type annotation.

Maybe something like "Fix partial evaluation of ForwardRef objects that rely on names defined as globals."

dr-carlos replied:
Perfect, thanks!

JelleZijlstra commented on Lib/test/test_annotationlib.py:
```diff
@@ -1675,6 +1675,24 @@ def test_name_lookup_without_eval(self):
         self.assertEqual(exc.exception.name, "doesntexist")
 
+    def test_evaluate_undefined_generic(self):
+        # Test the codepath where have to eval() with undefined variables.
```
Can you use a test case that doesn't rely on direct construction of ForwardRef objects? The ForwardRef constructor is undocumented.

dr-carlos replied:
Yep, can do! I was using the constructor since the other `ForwardRef` tests also tend to use it, but I'll just rewrite it to use classes instead (as in the issue).

dr-carlos replied:
> Can you use a test case that doesn't rely on direct construction of ForwardRef objects? The ForwardRef constructor is undocumented.

I've also removed the builtin overriding globals test as I couldn't figure out a way to test it without the `ForwardRef` constructor, since `get_annotations()` evaluates the builtins before they are passed to this code path.

dr-carlos committed:
Use `get_annotations()` to access annotations, check all args in undefined generic tests

Removed global overriding builtin test as it seemingly wasn't compatible with `get_annotations()`.

dr-carlos committed:
Update NEWS entry to more broadly reflect change

dr-carlos committed:
Fix typo in NEWS entry
